global-api/importer/crosswalk/utils.py

```diff
@@ -0,0 +1,532 @@
+    """
+    xlon, ylat = meshgrid(lons, lats)
+    R = earth_radius(ylat)
+
+    dlat = deg2rad(gradient(ylat, axis=0))
+    dlon = deg2rad(gradient(xlon, axis=1))
+
+    dy = dlat * R
+    dx = dlon * R * cos(deg2rad(ylat))
+
+    area = dy * dx
+
+    xda = DataArray(
+        area,
+        dims=["lat", "lon"],
+        coords={"lat": lats, "lon": lons},
+        attrs={
+            "long_name": "area_per_pixel",
+            "description": "area per pixel",
+            "units": "m^2",
+        },
+    )
+    return xda
+
+
+def locode_to_shapefile(session, locode):
+    """get shapefile from locode"""
+    query = text("""SELECT geometry FROM osm WHERE locode = :locode;""")
+    return session.execute(query, {"locode": locode}).fetchall()
+
+
+def shp_to_gpd(shp, crs="EPSG:4326"):
+    """convert list of shapes to geopandas dataframe"""
+    return gpd.GeoDataFrame(
+        geometry=[shapely.wkt.loads(geom) for geom, in shp]
+    ).set_crs(crs)
+
+
+def shp_to_geojson(shp):
+    """convert SHP to geoJSON and returns geometries"""
+    geometry = loads(shp[0][0])
+    geojson_dict = mapping(geometry)
+    geojson_string = json.dumps(geojson_dict)
+    return [geojson.loads(geojson_string)]
+
+
+def bounds_from_gdf(gdf):
+    """(west, south, east, north) bounds from geopandas dataframe"""
+    return gdf.geometry.unary_union.bounds
+
+
+def bounds_from_polygon(polygon):
+    """(west, south, east, north) from shapely polygon"""
+    return polygon.bounds
+
+
+def xarray_to_gdf(ds):
+    lats = ds["lat"].values
+    lons = ds["lon"].values
+    x_res, y_res = ds.rio.resolution()
+    x_res_half = x_res / 2
+    y_res_half = y_res / 2
+
+    polygons = []
+
+    for lon in lons:
+        for lat in lats:
+            left = lon - x_res_half
+            right = lon + x_res_half
+            top = lat + y_res_half
+            bottom = lat - y_res_half
+
+            polygon = Polygon(
+                [(left, bottom), (right, bottom), (right, top), (left, top)]
+            )
+            polygons.append(polygon)
+
+    return gpd.GeoDataFrame(geometry=polygons, crs="EPSG:4326")
+
+
+def gdf_to_xarray(gdf):
+    lons = gdf.to_crs(4326).centroid.x.round(2)
+    lats = gdf.to_crs(4326).centroid.y.round(2)
+
+    lats_unique = list(set(sorted(lats)))
+    lons_unique = list(set(sorted(lons)))
+
+    lon_coord = xr.DataArray(lons_unique, dims="lon", name="lon")
+    lat_coord = xr.DataArray(lats_unique, dims="lat", name="lat")
+
+    return xr.Dataset(
+        {
+            "lat": lat_coord,
+            "lon": lon_coord,
+        }
+    ).sortby(["lon", "lat"])
+
+
+def xarray_intersection(
+    ds,
+    geometries,
+    all_touched=True,
+    crs=4326,
+    x_dim="lon",
+    y_dim="lat",
+    *args,
+    **kwargs,
+):
+    """previously called get_intersection"""
+    return ds.rio.set_spatial_dims(x_dim=x_dim, y_dim=y_dim).rio.clip(
+        geometries, crs=crs, all_touched=all_touched
+    )
+    
```

Comment:
This works. However, since the data is hosted on a THREDDS data server (TDS) we could use unidata's [siphon package](https://github.com/Unidata/siphon). Below is some code showing how to access crosswalk data using siphon

```python
from siphon.catalog import TDSCatalog
import xarray as xr

def tds_catalog(url):
    """create a catalog client"""
    return TDSCatalog(url)


def tds_generator(catalog):
    """generator for files in a THREDDS catalog"""
    yield from (dataset for dataset in catalog.datasets)
        

def get_dataset_url(catalog, dataset):
    """retrieve the URL to the dataset"""
    return catalog.datasets[dataset].access_urls["OPENDAP"]


# ===================================
# example
# ===================================

# create TDS catalog
url = "https://thredds.daac.ornl.gov/thredds/catalog/ornldaac/1741/catalog.xml"
catalog = tds_catalog(url)

# lists all the datasets
gen = tds_generator(catalog)
datasets = [fl for fl in gen]

# filter the datasets by keywords
keywords = ['US', 'airport']
datasets_filtered = [dataset for dataset in datasets if dataset.endswith('_hi.nc4') and all(keyword in dataset for keyword in keywords)]

# read dataset_url
dataset_url = get_dataset_url(catalog, datasets_filtered[0])
ds = xr.open_dataset(dataset_url)
```